modules/ethereum/src/lib.rs

```diff
@@ -179,18 +187,23 @@ impl ImportContext {
 ///
 /// Storage modification must be discarded if block import has failed.
 pub trait Storage {
+	/// Header submitter identifier.
+	type Submitter: Clone + Ord;
+
 	/// Get best known block.
 	fn best_block(&self) -> (u64, H256, U256);
 	/// Get last finalized block.
 	fn finalized_block(&self) -> (u64, H256);
 	/// Get imported header by its hash.
-	fn header(&self, hash: &H256) -> Option<Header>;
+	///
+	/// Returns header and its submitter (if known).
+	fn header(&self, hash: &H256) -> Option<(Header, Option<Self::Submitter>)>;
 	/// Get header import context by parent header hash.
 	fn import_context(&self, parent_hash: &H256) -> Option<ImportContext>;
 	/// Get new validators that are scheduled by given header.
 	fn scheduled_change(&self, hash: &H256) -> Option<Vec<Address>>;
 	/// Insert imported header.
-	fn insert_header(&mut self, header: HeaderToImport);
+	fn insert_header(&mut self, submitter: Option<Self::Submitter>, header: HeaderToImport);
```

Comment:
Why isn't `submitter` part of `HeaderToImport`? I feel that this would look a bit nicer.